src/opencl/mod.rs

```diff
@@ -0,0 +1,158 @@
+            inout: *mut u8,
+            len: usize,
+            iv: *const u8,
+            layers: usize,
+            instances: *const u8,
+        ) -> i32;
+
+        pub(super) fn sloth256_189_opencl_init(
+            error: &mut i32,
+        ) -> *const u8;
+
+        pub(super) fn sloth256_189_pinned_alloc(
+            instances: *const u8,
+            size: usize,
+            error: &mut i32,
+        ) -> *mut u8;
+
+        pub(super) fn sloth256_189_pinned_free(
+            instances: *const u8,
+        ) -> i32;
+
+        pub(super) fn sloth256_189_opencl_cleanup(
+            instances: *const u8,
+        ) -> i32;
+    }
+}
+
+/// Allocate pinned and aligned host memory. This makes memory copy operations on NVIDIA GPUs faster
+/// and allignment is required for zero-copy buffers for Intel integrated graphics devices
+/// Call this function after calling the init function
+pub fn pinned_memory_alloc(instance: *const u8, size: usize) -> Result<Vec<u8>, OpenCLEncodeError> {
+    let mut return_code: i32 = 0;
+
+    let pointer = unsafe {
+        ffi::sloth256_189_pinned_alloc(instance, size, &mut return_code)
+    };
+
+    if return_code != 0 {
+        return Err(OpenCLEncodeError::OpenCLError(return_code));
+    }
+
+    let pointer_as_vec = unsafe{
+        Vec::from_raw_parts(pointer, size, size)
+    };
+
+    return Ok(pointer_as_vec);
+}
+
+/// Free the pinned memory allocated with pinned_memory_alloc function above. A std::mem::forget({vector})
+/// call is required since the vector in Rust will be using memory freed in C.
+/// Call this function before calling the cleanup function
+pub fn pinned_memory_free(instances: *const u8) -> Result<(), OpenCLEncodeError> {
+    let return_code = unsafe {
+        ffi::sloth256_189_pinned_free(instances)
+    };
+
+    if return_code != 0 {
+        return Err(OpenCLEncodeError::OpenCLError(return_code));
+    }
+
+    return Ok(());
+}
+
+/// Initializes the encode kernel for a given device. In essence performs the tasks that must be done only once at the start.
+/// Returns a pointer to the initialized instance which should be passed to the encode and cleanup functions.
+/// Calling this once at the start is sufficient.
+pub fn initialize() -> Result<*const u8, OpenCLEncodeError> {
+    let mut return_code: i32 = 0;
+    let instances = unsafe {
+        ffi::sloth256_189_opencl_init(&mut return_code)
+    };
+
+    if return_code != 0 {
+        return Err(OpenCLEncodeError::OpenCLError(return_code));
+    }
+
+    return Ok(instances);
+}
+
+/// Cleans up the resources allocated in the initialization of the encode kernel.
+/// Calling this once at the end is sufficient.
+pub fn cleanup(instances: *const u8) -> Result<(), OpenCLEncodeError> {
+    let return_code = unsafe {
+        ffi::sloth256_189_opencl_cleanup(instances)
+    };
+
+    if return_code != 0 {
+        return Err(OpenCLEncodeError::OpenCLError(return_code));
+    }
+
+    return Ok(());
+}

```

Comment:
Functions like these are inherently unsafe.

All of these functions should either be private and re-exposed as one (if there is no reason to not use pinned memory for instance) or wrapped into a struct with public API that is impossible to use incorrectly.

Reply:
That seems like a reasonable approach (creating a struct with a public API). Merging them all into a single function would make it so that initialization and cleanup is done multiple times, which is a major performance loss for data of smaller sizes.